### PR TITLE
Suppress redundant queries in StudentHomePageAction, part 1

### PR DESCRIPTION
**Pull request by wkurniawan07**
Part of #3650

### CODE REVIEW

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 1]
```diff
@@ -153,4 +165,14 @@ private Course getCourseEntity(String courseId) {
...
+    
+    private List<Course> getCourseEntities(List<String> courseIds) {
+        Query q = getPM().newQuery(Course.class);
+        q.setFilter(":p.contains(ID)");
+
+        @SuppressWarnings("unchecked")
+        List<Course> courses = (List<Course>) q.execute(courseIds);
+
+        return courses;
+    }
```
Comment (kanghj) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Might not need to change this, I think this still makes the same number of queries as the original?

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 2]
```diff
@@ -139,7 +140,8 @@ public CourseDetailsBundle getCourseDetails(String courseId) throws EntityDoesNo
         return courseSummary;
     }
 
-    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId) 
+    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId,
+                                                                    Map<String, StudentAttributes> ctaMap)
```
Comment (damithc) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: The method signature is no longer self-explanatory. Need a header comment.

Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Added some comments

File: src/main/java/teammates/storage/api/CoursesDb.java (hunk 1 above)
Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/CoursesDb.java]: batchquery
Top = querying one by one, bottom = batch query.
Batch query saves 140 micropennies per course but adding 70 micropennies per course to the batch get. Overall, still saving 70 micropennies per course.

**Comment by wkurniawan07** [comment]
Currently, with 3 courses and 3 sessions in each course, managed to reduce the cost of loading StudentHomePage from 13580 to 11480

**Comment by wkurniawan07** [comment]
Currently, with 3 courses and 3 sessions in each course, managed to reduce the cost of loading StudentHomePage from 13580 to 8260

File: src/main/java/teammates/logic/api/Logic.java [hunk 3]
```diff
@@ -785,16 +786,20 @@ public static boolean isCourseArchived(CourseAttributes course, InstructorAttrib
...
+     * The second parameter ctaMap is used to keep the query results in the form of
+     * course ID -> {@link StudentAttributes} mapping so that they can be re-used.
      * 
      * @return Details of courses the student is in. CourseData objects
      *         returned contain details of feedback sessions too (except the ones
      *         still AWAITING).
      */
-    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId)
+    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId,
+                                                                    Map<String, StudentAttributes> ctaMap)
```
Comment (damithc) [inline src/main/java/teammates/logic/api/Logic.java]: `ctaMap` rename to `alreadyRetrievedData` or something similar?

File: src/main/java/teammates/logic/api/Logic.java [hunk 4]
```diff
@@ -785,16 +786,20 @@ public static boolean isCourseArchived(CourseAttributes course, InstructorAttrib
     /**
      * Preconditions: <br>
      * * All parameters are non-null.
+     * <br>
+     * The second parameter ctaMap is used to keep the query results in the form of
```
Comment (damithc) [inline src/main/java/teammates/logic/api/Logic.java]: Use the usual javadoc format for parameter description. No need to say `The second parameter...`

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 5]
```diff
@@ -148,6 +154,10 @@ public CourseDetailsBundle getCourseDetails(String courseId) throws EntityDoesNo
         for (CourseAttributes c : courseList) {
 
             StudentAttributes s = studentsLogic.getStudentForCourseIdAndGoogleId(c.id, googleId);
+            if (ctaMap != null && ctaMap instanceof Map<?, ?>) {
+                // the map is null in logic test
```
Comment (damithc) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: It's best if production code doesn't refer to test code.

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java [hunk 6]
```diff
@@ -26,9 +26,14 @@ public ActionResult execute() throws EntityDoesNotExistException {
...
-    
-            data.sessionSubmissionStatusMap = generateFeedbackSessionSubmissionStatusMap(data.courses, account.googleId);
+            Map<String, StudentAttributes> courseToAttributesMap = new HashMap<String, StudentAttributes>();
+            data.courses = logic.getCourseDetailsListForStudent(account.googleId, courseToAttributesMap);
+            /*
+             * courseToAttributesMap is used to keep the query results in the form of
+             * course ID -> StudentAttributes mapping so that they can be re-used.
+             */
+            data.sessionSubmissionStatusMap = generateFeedbackSessionSubmissionStatusMap(data.courses, account.googleId,
+                                                                                         courseToAttributesMap);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: `courseToAttributes` is ambiguous. What `Attributes`? May be use `courseStudentMap`?

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java [hunk 7]
```diff
@@ -58,27 +63,33 @@ public ActionResult execute() throws EntityDoesNotExistException {
...
 
-    private boolean getStudentStatusForSession(FeedbackSessionAttributes fs, String googleId){
-        StudentAttributes student = logic.getStudentForGoogleId(fs.courseId, googleId);
+    private boolean getStudentStatusForSession(FeedbackSessionAttributes fs, String googleId,
+                                               Map<String, StudentAttributes> ctaMap) {
+        StudentAttributes student = ctaMap.get(fs.courseId);
+        if (student == null) {
+            // this is a fail-safe branch that normally will not be reached
+            student = logic.getStudentForGoogleId(fs.courseId, googleId);
+        }
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: It is not good to have this kind of 'fail safe' branches. Either it can be null or it cannot be null. If it cannot be, we can put an assertion.

**Comment by damithc** [comment]
Branch name doesn't follow the standard.

**Comment by wkurniawan07** [comment]
Just like refinements, this is not for any issue per se. Is there a branch name standard for that?

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 5 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: So... just remove this comment?

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java (hunk 6 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: How about `courseToStudentAttributes`?

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java (hunk 7 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: I can remove it. It's not covered in tests, just as expected.

File: src/main/java/teammates/logic/api/Logic.java (hunk 3 above)
Comment (damithc) [inline src/main/java/teammates/logic/api/Logic.java]: After more examination, it looks like you are using an input parameter to retrieve output? That's 'sneaky'. I'd rather not do 'sneaky' stuff in the code. Having a return value adds to the confusion. May be define a new return data type that contains all the values you need?

**Comment by wkurniawan07** [comment]
Did improvements as suggested.

File: src/main/java/teammates/logic/api/Logic.java [hunk 8]
```diff
@@ -1587,15 +1592,13 @@ public String getFeedbackSessionResultSummaryInSectionAsCsv(String courseId,
      * Preconditions: <br>
      * * All parameters are non-null.
      */
-    public boolean hasStudentSubmittedFeedback(String courseId, String feedbackSessionName, String studentEmail)
+    public boolean hasStudentSubmittedFeedback(FeedbackSessionAttributes fsa, String studentEmail)
             throws InvalidParametersException, EntityDoesNotExistException {
```
Comment (damithc) [inline src/main/java/teammates/logic/api/Logic.java]: Is this change related to this PR? If not, should do as a separate issue. On a related note, passing an entire object is not necessarily better.  It increases coupling unnecessarily. A method should not receive more than it needs.

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 5 above)
Comment (damithc) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Fix the test rather than write production code to cater for a test.

File: src/main/java/teammates/logic/api/Logic.java (hunk 8 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/Logic.java]: Yes. Under the old one, there is a repeated query of FeedbackSessionAttributes with the courseId and feedbackSessionName taken from an existing Attributes from a query result, which is redundant.
And it's not being replaced (replacing introduces regression, I've experienced it), I just add another method that takes in the entire Attributes and process it.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 9]
```diff
@@ -143,12 +143,24 @@ public FeedbackSessionAttributes copyFeedbackSession(String newFeedbackSessionNa
...
         }
+        return getFeedbackSessionsForUserInCourseSkipCheck(courseId, userEmail);
+    }
 
+    /**
+     * @param courseId
+     * @param userEmail
+     * @return a list of viewable feedback sessions for any user for his course.
+     * @throws EntityDoesNotExistException
+     */
```
Comment (damithc) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: The comment doesn't mention anything about the 'skipTest' part. In any case, I try to avoid this kind of method variations. They are a sign that the code is getting messy.

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 10]
```diff
@@ -155,7 +155,7 @@ public StudentAttributes getStudentForEmail(String courseId, String email) {
      * <br> * All parameters are non-null.
      * @return null if no such student is found. 
      */
-    public StudentAttributes getStudentForGoogleId(String courseId, String googleId) {
+    public StudentAttributes getStudentForCourseIdAndGoogleId(String courseId, String googleId) {
```
Comment (damithc) [inline src/main/java/teammates/storage/api/StudentsDb.java]: Is this relevant to this PR?

**Comment by damithc** [comment]
Can break this down to smaller PRs? I'm not sure if the complexities introduced are worth the performance gain or if there is a simpler way to do it. I'm unable to judge because the PR is too big.

**Comment by damithc** [comment]
Re PR name, it's fine.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java (hunk 9 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: The way I write it, the old method `getFeedbackSessionsForUserInCourse` calls the skipCheck method after verifying that the course is present. In one way, it can be seen as an afterCheck method. But also it can be directly used, i.e a skipCheck method.
The rationale for writing the skipCheck version is the courseIds passed are obtained from CourseAttributes resulting from query, so checking presence of those courses are redundant.

File: src/main/java/teammates/logic/api/Logic.java (hunk 8 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/Logic.java]: Wait, I'm not exactly talking about the right stuff. The one I talked about above is the method that follows, i.e `isFeedbackSessionCompletedByStudent`. Still, this Logic method is changed as an (successful) attempt to reduce redundant queries.

**Comment by wkurniawan07** [comment]
I will interactive-rebase this PR first to group the commits according to areas of improvement. If the PR is still too big, I should be able to remove some groups.
afterimprovement
Anyway, here's the improvement. Ran StudentHomePageUiTest once on each. Top is the current master, bottom is current PR.

**Comment by wkurniawan07** [comment]
Interactive rebase done. Now the 5 existing commits are independent of each other and each targets different area of improvement. If we're to separate this PR, which ones to include in which PR is just a matter of choice.

**Comment by damithc** [comment]
Nice. Yes, create separate PRs. I don't even mind five different PRs (smaller the better). The idea is to make each one as small as possible but also independent and justifiable on its own.

**Comment by wkurniawan07** [comment]
Dr @damithc you can refer to #3667 #3668 #3669 and #3670. I'll reset this PR to only include the first commit.
Explanation for the first commit: batch query is faster than querying one by one.

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 11]
```diff
@@ -153,4 +165,14 @@ private Course getCourseEntity(String courseId) {
     
         return courseList.get(0);
     }
+    
+    private List<Course> getCourseEntities(List<String> courseIds) {
+        Query q = getPM().newQuery(Course.class);
+        q.setFilter(":p.contains(ID)");
```
Comment (damithc) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Can you add a comment here, pointing to the documentation of this filter?

Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Hmmm the jar has no attached Javadoc, so not sure what I should point it to. Also, this type of filter (`:p.contains(paramName)`) has already been used in a few other Dbs.

Comment (damithc) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Ok. No need to add a comment.

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 12]
```diff
@@ -65,6 +65,18 @@ public CourseAttributes getCourse(String courseId) {
         return new CourseAttributes(c);
     }
     
+    public List<CourseAttributes> getCourses(List<String> courseIds) {
+        Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, courseIds);
+        List<Course> courses = getCourseEntities(courseIds);
+        List<CourseAttributes> courseAttributes = new ArrayList<CourseAttributes>();
+        for (Course c: courses) {
+            if (c != null) {
```
Comment (damithc) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Can c be null?

Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Oh right. It won't be null.

Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Removed the null check.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 13]
```diff
@@ -463,16 +463,12 @@ public CourseSummaryBundle getCourseSummaryWithoutStats(String courseId) throws
             throw new EntityDoesNotExistException("Student with Google ID " + googleId + " does not exist");
         }
         
-        ArrayList<CourseAttributes> courseList = new ArrayList<CourseAttributes>();
-
+        List<String> courses = new ArrayList<String>();
```
Comment (damithc) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Rename to `courseIDs`?

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 14]
```diff
@@ -65,6 +65,16 @@ public CourseAttributes getCourse(String courseId) {
...
     }
     
+    public List<CourseAttributes> getCourses(List<String> courseIds) {
+        Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, courseIds);
+        List<Course> courses = getCourseEntities(courseIds);
+        List<CourseAttributes> courseAttributes = new ArrayList<CourseAttributes>();
+        for (Course c: courses) {
+            courseAttributes.add(new CourseAttributes(c));
+        }
+        return courseAttributes;
```
Comment (damithc) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Add a TODO here to extract a method `CourseAttributes.convert(List<Course>): List<CourseAttributes>`

**Comment by damithc** [comment]
Can merge after addressing the minor comments.
